### PR TITLE
CLI: allow to download templates in sb repro

### PR DESCRIPTION
**Pull request by yannbf**
Issue: N/A

## What I did

This PR integrates `degit` into the sb cli.

It adds two features:

### Prefer downloading from storybook repro repository

This way, the CLI just downloads a git repo from [repro-templates](https://github.com/storybookjs/repro-templates) instead of having to run a generator (e.g. create-react-app myApp...). This could save time both in CI but also for the `sb repro` experience.

If the selected template does not exist in repro-templates, it will fallback to the original generator.

For the user, there's no change in the command. Optionally, you could use the `--skip-git` flag to use the generators directly instead.

The repro template has both `main` and `next` branches which old latest vs next versions of Storybook. The script will detect whether the user is using `sb repro` or `sb repro@some-alpha-version` and will download from the correct branch

### Set project via --url flag

`sb repro` now has a `--url` flag which allows you to download any git project (github, gitlab, bitbucket..) and bootstrap Storybook on top of it. Could be useful for QAing purposes in case we find big projects with certain settings that could be valuable to test out. 

This uses `degit` under the hood so you can use all kinds of formats for the URL such as `username/project` or `https://github.com/username/project` or other formats that degit supports.

## How to test

First check the branch, bootstrap it or just build the cli:
```
yarn build cli
```

Next, somewhere in an experimentation folder, run `sb repro` using the URL of a project that does not contain Storybook.

```
# This will prompt for location folder, using the repo as default folder name
../storybook/lib/cli/bin/index.js repro --url the-url-here
```

```
# This will download and install things automatically with no prompt
../storybook/lib/cli/bin/index.js repro --url the-url-here outputfoldername
```

- [ ] Is this testable with Jest or Chromatic screenshots?
- [ ] Does this need a new example in the kitchen sink apps?
- [ ] Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

### CODE REVIEW

File: code/lib/cli/src/repro-generators/scripts.ts [hunk 1]
```diff
@@ -287,8 +291,24 @@ export const createAndInit = async (
...
+    await doTask(generate, {
+      ...options,
+      generator: options.gitRepoGenerator,
+      cwd: options.creationPath,
+    });
+  } catch (err) {
+    // fallback to local generators
+    logger.info(`⏩ Skipping git template and using local generator instead`);
+    await doTask(generate, { ...options, cwd: options.creationPath });
+  }
```
Comment (tmeasday) [inline code/lib/cli/src/repro-generators/scripts.ts:308]: I find this control flow via `throw` to be pretty confusing. My suggestion would be something like:

```ts
let { preferGitTemplate } = options;
try {
  if (preferGitTemplate) await doTask(...);
} catch(err) {
  logger.info(...);
  preferGitTemplate = false;
}

if (!preferGitTemplate) await doTask(...);
```

Comment (tmeasday) [reply code/lib/cli/src/repro-generators/scripts.ts:308]: Or even better restructure to something like

```ts
try {
  if (options.preferGitTemplate) return doTask(...)
} catch (err) {
   logger.info(...);
}
return doTask(...);
```

**Review by tmeasday — APPROVED** [review]
LGTM! One suggestion that might be a bit easier to read

**Comment by yannbf** [comment]
Given the state of things now, after the whole rebasing and the new idea for our E2E setup, this PR should not be merged.
I will create a new command that works solely based on --url and using degit from the new repro templates format.

**Pull request closed without merging**